src/messageschannel.cpp

```diff
@@ -90,14 +97,64 @@ MatrixMessagesChannel::MatrixMessagesChannel(MatrixConnection *connection, Tp::B
     }
 }
 
+void MatrixMessagesChannel::onPendingEventChanged(int pendingEventIndex)
+{
+    // Delivery Report message
+    // https://telepathy.freedesktop.org/spec/Channel_Interface_Messages.html#Enum:Delivery_Status
+    // https://matrix.org/docs/spec/client_server/r0.4.0.html#put-matrix-client-r0-rooms-roomid-send-eventtype-txnid
+
+    const QMatrixClient::PendingEventItem &pendingEvent = m_room->pendingEvents().at(pendingEventIndex);
+    Tp::DeliveryStatus tpDeliveryStatus;
+    switch (pendingEvent.deliveryStatus()) {
+    case QMatrixClient::EventStatus::ReachedServer:
+        tpDeliveryStatus = Tp::DeliveryStatusAccepted;
+        break;
+    case QMatrixClient::EventStatus::SendingFailed:
+        tpDeliveryStatus = Tp::DeliveryStatusTemporarilyFailed;
+        break;
+    default:
+        tpDeliveryStatus = Tp::DeliveryStatusUnknown;
+        break;
+    }
+
+    Tp::MessagePartList partList;
+
+    Tp::MessagePart header;
+    header[QStringLiteral("message-sender")]    = QDBusVariant(m_targetHandle);
+    header[QStringLiteral("message-sender-id")] = QDBusVariant(m_targetId);
+    header[QStringLiteral("message-type")]      = QDBusVariant(Tp::ChannelTextMessageTypeDeliveryReport);
+    header[QStringLiteral("delivery-status")]   = QDBusVariant(tpDeliveryStatus);
+    header[QStringLiteral("delivery-token")]    = QDBusVariant(pendingEvent.event()->transactionId());
+    partList << header;
+
+    addReceivedMessage(partList);
+}
+
 MatrixMessagesChannelPtr MatrixMessagesChannel::create(MatrixConnection *connection, Tp::BaseChannel *baseChannel)
 {
     return MatrixMessagesChannelPtr(new MatrixMessagesChannel(connection, baseChannel));
 }
 
 QString MatrixMessagesChannel::sendMessageCallback(const Tp::MessagePartList &messageParts, uint flags, Tp::DBusError *error)
 {
-    return QString();
+    if (!m_room) {
+        // TODO: error?
+        qDebug() << Q_FUNC_INFO;
+        return QString();
+    }
+    QString content;
+    foreach (const Tp::MessagePart &part, messageParts) {
+        if (part.contains(QStringLiteral("content-type"))
+                && part.value(QStringLiteral("content-type")).variant().toString() == QStringLiteral("text/plain")
+                && part.contains(QStringLiteral("content"))) {
+            content = part.value(QStringLiteral("content")).variant().toString();
+            break;
+        }
+    }
+
+    connect(m_room, &QMatrixClient::Room::pendingEventChanged, this, &MatrixMessagesChannel::onPendingEventChanged);
+    auto txnId = m_room->postPlainText(content);
```

Comment:
According to [Qt style](https://wiki.qt.io/Coding_Conventions#auto_Keyword), the 'auto' keyword is allowed only in two cases (and this one is not appreciated). Please use the right type instead of 'auto'.